### PR TITLE
Lesson3/3

### PR DESCRIPTION
**Pull request by Natasha-eng**

### CODE REVIEW

File: sources/views/contacts.js [hunk 1]
```diff
@@ -71,20 +73,23 @@ export default class Contacts extends JetView {
...
-		this.on(contacts.data, "onStoreUpdated", (id, obj, mode) => {
+		var contactsdDp = webix.dp(contacts);
 
-			if (mode === "add") {
-				contactsList.select(id);
-			} else if (mode === "delete") {
-				const firstId = contactsList.getFirstId();
-				contactsList.select(firstId);
-			}
+		contactsdDp.attachEvent("onAfterSave", function (response) {
```
Comment (kullias) [inline sources/views/contacts.js]: you do not need this
you have waitSave with add. wait for the response and select the id that is returned

File: sources/views/contacts.js [hunk 2]
```diff
@@ -43,11 +45,12 @@ export default class Contacts extends JetView {
...
 									"Name": "",
 									"Email": "",
-									"Status": 1,
-									"Country": 1
+									"Status": 4,
+									"Country": 5
 								};
-
-								contacts.add(newContact);
+								contacts.waitSave(() => {
```
Comment (kullias) [inline sources/views/contacts.js]: here you need to wait for the ID from the server and then select. use the promise returned by waitSave

File: sources/views/contacts.js [hunk 3]
```diff
@@ -14,6 +15,7 @@ export default class Contacts extends JetView {
 					rows: [
 						{
 							view: "list",
+							save: saveContacts,
```
Comment (kullias) [inline sources/views/contacts.js]: you do not need this

File: sources/models/records.js [hunk 4]
```diff
@@ -1,8 +1,29 @@
-export const data = new webix.DataCollection({ data:[
-	{ id:1, title:"The Shawshank Redemption", year:1994, votes:678790, rating:9.2, rank:1},
-	{ id:2, title:"The Godfather", year:1972, votes:511495, rating:9.2, rank:2},
-	{ id:3, title:"The Godfather: Part II", year:1974, votes:319352, rating:9.0, rank:3},
-	{ id:4, title:"The Good, the Bad and the Ugly", year:1966, votes:213030, rating:8.9, rank:4},
-	{ id:5, title:"My Fair Lady", year:1964, votes:533848, rating:8.9, rank:5},
-	{ id:6, title:"12 Angry Men", year:1957, votes:164558, rating:8.9, rank:6}
-]});
\ No newline at end of file
+function saveContacts(id, operation) {
```
Comment (kullias) [inline sources/models/records.js]: you do not need this file. the only connection between the client side and the backend side - the models with dataprocessors. save: "rest->url" and that's it

File: sources/views/countries.js [hunk 5]
```diff
@@ -1,10 +1,12 @@
 import { countries } from "../models/countries";
+import { saveCountries } from "../models/records";
 import BaseTable from "./baseTable";
 
 export default class CountriesView extends BaseTable {
   constructor(app) {
     super(app, {
       collection: countries,
+      save: saveCountries,
```
Comment (kullias) [inline sources/views/countries.js]: you do not need this

File: sources/views/form.js [hunk 6]
```diff
@@ -95,12 +94,18 @@ export default class FormView extends JetView {
         const form = this.$$("contactForm");
         const id = url[0].params.id;
 
-        if (id) {
-            const contact = contacts.getItem(id);
-            form.setValues(contact);
-        } else {
-            form.clear();
-        }
+        webix.promise.all([
```
Comment (kullias) [inline sources/views/form.js]: you do not need for this data every time in url change
you need to wait for the data 1 time
in config()
config() can return a promise

```js
config() {
        const _ = this.app.getService("locale")._;
        return webix.promise.all([
                contacts.waitData,
                statuses.waitData,
                countries.waitData
            ]).then(() => {
                return {
                view: "form",
               ...
```

File: sources/views/statuses.js [hunk 7]
```diff
@@ -1,10 +1,12 @@
+import { saveStatuses } from "../models/records";
 import { statuses } from "../models/statuses";
 import BaseTable from "./baseTable";
 
 export default class StatusesView extends BaseTable {
     constructor(app) {
         super(app, {
             collection: statuses,
+            save: saveStatuses,
```
Comment (kullias) [inline sources/views/statuses.js]: you do not need this

**Review by kullias — CHANGES_REQUESTED** [review]
there is a minor issue with templates:
- add something without data
- refresh the page
- you will have undefined in the list and the tables

solution: check if the string exists. e.g. template: obj => `${obj.Name || ""} ${obj.Email || ""}`

File: sources/views/contacts.js [hunk 8]
```diff
@@ -15,23 +14,28 @@ export default class Contacts extends JetView {
...
 
 							onClick: {
 								removeBtn: function (ev, id) {
-									contacts.remove(id);
-									return false;
+									contacts.waitSave(() => {
+										contacts.remove(id);
+										return false;
+									}).then(() => {
+										const firstId = contacts.getFirstId();
```
Comment (kullias) [inline sources/views/contacts.js]: if (firstId)

File: sources/views/form.js [hunk 9]
```diff
@@ -7,83 +7,88 @@ import { contacts } from "../models/contacts";
...
+                            data: countries
+                        }
+                    },
+                    {
+                        cols: [
+                            {
+                                view: "button",
+                                value: _("Save"),
+                                css: "webix_primary",
+                                click: () => {
```
Comment (kullias) [inline sources/views/form.js]: now it's time for some code style perfectionism
the code of config() got long, so you should move handlers to class methods. e.g. click: () => this.SaveForm(), where SaveForm() contains the code of this click handler

File: sources/views/form.js [hunk 10]
```diff
@@ -7,83 +7,88 @@ import { contacts } from "../models/contacts";
...
+                                        webix.message("Contact is updated.");
+                                    } else {
+                                        webix.message("Contact hasn't been changed.");
+                                    }
+                                }
+                            },
+                            {
+                                view: "button",
+                                value: _("Cancel"),
+                                click: () => {
```
Comment (kullias) [inline sources/views/form.js]: you should also move out this handler

File: sources/views/form.js (hunk 9 above)
Comment (kullias) [reply sources/views/form.js]: mb this is also true for other view classes

File: sources/views/contacts.js [hunk 11]
```diff
@@ -27,38 +28,15 @@ export default class Contacts extends JetView {
...
-									contacts.waitSave(() => {
-										contacts.remove(id);
-										return false;
-									}).then(() => {
-										const firstId = contacts.getFirstId();
-										this.select(firstId)
-									})
-
-								},
+								removeBtn: (ev,id) => this.removeContact(ev,id),
```
Comment (kullias) [inline sources/views/contacts.js]: ev is not used, do not pass it to the method

File: sources/views/contacts.js [hunk 12]
```diff
@@ -86,6 +64,32 @@ export default class Contacts extends JetView {
...
 			if (firstId)
 				this.contactsList.select(firstId);
-		})
+		});
+	}
+
+	removeContact(ev,id) {
+		contacts.waitSave(() => {
+			contacts.remove(id);
+			return false;
```
Comment (kullias) [inline sources/views/contacts.js]: I missed it. return false should be outside of promise, at the end of the method

**Review by kullias — APPROVED** [review]

**Review by HelgaListopad — APPROVED** [review]

**Pull request merged**